drivers/dcf77/dcf77.c

```diff
@@ -0,0 +1,213 @@
+                dev->internal_state = DCF77_STATE_RX;
+            }
+            else {
+                dev->internal_state = DCF77_STATE_IDLE;
+            }
+            break;
+        case DCF77_STATE_RX:
+            DEBUG("[dcf77] EVENT RX 1 !\n");
+            dev->startTime = xtimer_now_usec();
+            break;
+    }
+}
+
+static void _level_cb_low(dcf77_t *dev)
+{
+    switch (dev->internal_state) {
+        case DCF77_STATE_IDLE:
+            DEBUG("[dcf77] EVENT IDLE 0  !\n");
+            dev->startTime = xtimer_now_usec();
+            dev->internal_state = DCF77_STATE_START;
+            break;
+        case DCF77_STATE_RX:
+            DEBUG("[dcf77] EVENT RX 0 !\n");
+            if ((xtimer_now_usec() - dev->startTime) >
+                DCF77_PULSE_WIDTH_THRESHOLD) {
+                dev->bitseq.bits |=  1ULL << dev->bitCounter;
+            }
+
+            dev->bitCounter++;
+
+            if (dev->bitCounter >= DCF77_READING_CYCLE) {
+                dev->bitCounter = 0;
+                dev->startTime = xtimer_now_usec();
+                dev->last_bitseq.bits = dev->bitseq.bits;
+                dev->internal_state = DCF77_STATE_START;
+
+                if (dev->tick_cb) {
+                    dev->tick_cb(dev, dev->tick_cb_args);
+                }
+            }
+
+            break;
+    }
+}
+
+static void _level_cb(void *arg)
+{
+    dcf77_t *dev = (dcf77_t *)arg;
+
+    if (gpio_read(dev->params.pin)) {
+        _level_cb_high(dev);
+    }
+    else {
+        _level_cb_low(dev);
+    }
+}
+
+/**
+ * @brief   Initialize the Device
+ *
+ * @param   dev     device
+ * @param   params  device_params
+ *
+ * @retval  DCF77_OK             Success
+ * @retval  DCF77_INIT_ERROR     Error in initialization
+ */
+int dcf77_init(dcf77_t *dev, const dcf77_params_t *params)
+{
+    DEBUG("dcf77_init\n");
+
+    /* check parameters and configuration */
+    assert(dev && params);
+    dev->tick_cb = NULL;
+    dev->params = *params;
+    dev->internal_state = DCF77_STATE_IDLE;
+    dev->bitCounter = 0;
+    if (!gpio_init_int(dev->params.pin, dev->params.in_mode, GPIO_BOTH,
+                       _level_cb, dev)) {
+        return DCF77_OK;
+    }
+    else {
+        return DCF77_INIT_ERROR;
+    }
+}
+
+/**
+ * @brief   Formats the information after all bits of a cycle have been received
+ *
+ * @param   dev               device
+ * @param   time              datastruct for timeinformation
+ *
+ * @retval  DCF77_OK          Success
+ * @retval  DCF77_NOCSUM      Parity Bits aren't correct
+ */
+int dcf77_get_time(dcf77_t *dev, struct tm *time)
+{
+    assert(dev);
+
+    if (dev->last_bitseq.val.mesz == 2) {
+        time->tm_isdst = 1;
+    }
+    else {
+        time->tm_isdst = 0;
+    }
+
+    uint8_t minute = 10 * dev->last_bitseq.val.minute_h +
+                     dev->last_bitseq.val.minute_l;
+    if (__builtin_parity((dev->last_bitseq.bits >> DCF77_MINUTE_SHIFT) &
```

Comment:
This works for platforms?

Reply:
All but msp430 - but that should be resolved with #12457